packages/protocol/contracts/layer1/shasta/offchain-specs/BlockCalls.py

```diff
@@ -0,0 +1,287 @@
+from typing import Tuple, cast, List
+from eth_typing import Address
+import hashlib
+from eth_account import Account
+from eth_utils import keccak
+from Types import ProtoState, BlockArgs, Proposal, Content
+
+
+class BlockCalls:
+    """
+    Handles system calls that execute at the beginning and end of block processing.
+    These calls do not consume gas and are used for protocol-level operations.
+    TODOs:
+    - [ ] how to make sure there there is at least one non-zero anchor block hash?
+    """
+
+    DEFAULT_GAS_ISSUANCE_PER_SECOND = 1_000_000
+    GAS_ISSUANCE_PER_SECOND_MAX_OFFSET = 101
+    GAS_ISSUANCE_PER_SECOND_MIN_OFFSET = 99
+    ADDRESS_ZERO = cast(Address, "0x0000000000000000000000000000000000000000")
+    BLOCK_GAS_LIMIT = 100_000_000
+    TAIKO_DAO_TREASURE = cast(Address, "0x0000000000000000000000000000000000000123")
+    MAX_ANCHOR_BLOCK_HEIGHT_OFFSET = 64
+
+    def block_head_call(
+        self,
+        # provable: -> parent_block_hash
+        state: ProtoState,
+        # provable: -> proposal_hash
+        proposal: Proposal,
+        # provable: -> proposal_hash
+        proposal_content: Content,
+        # provable: -> content -> proposal_hash
+        block_args: BlockArgs,
+        # provable: -> parent_block_number -> parent_block_hash
+        block_number: int,
+        # provable: -> parent_block_hash
+        parent_prev_randao: str,
+        # provable: -> parent_block_hash
+        parent_timestamp: int,
+        # provable: -> up to 128 ancester block headers -> anchor_block_hash
+        anchor_block_hash: str,
```

Comment:
How do we expect the validity of the `anchor_block_hash` to be checked? I expected that, for example, we:
- We call `blockhash(proposedBlockNumber)` and store the result in `struct Proposal`.
- At proof generation time, we pass the `blockhashes` of required L1 blocks as private input, and check that they form a chain up to `blockhash(proposedBlockNumber)` .

But looks like we never call `blockhash` in inbox contract, so maybe you had something else in mind?